tests_e2e/playwright/navigator.py

```diff
@@ -0,0 +1,448 @@
+
+    def set_environment(self, kernel):
+        """Set environment of a jupyter notebook.
+
+        IMPORTANT: The focus MUST be on the notebook on which you want to set
+        the environment.
+
+        Conda environments may still be being built shortly after deployment.
+
+        Parameters
+        ----------
+        kernel: str
+            Name of kernel to set.
+
+        Returns
+        -------
+        None
+        """
+
+        popup = self._check_for_kernel_popup()
+        # if there is not a kernel popup, make it appear
+        if not popup:
+            self.page.get_by_text("Kernel", exact=True).click()
+            self.page.get_by_role("menuitem", name="Change Kernel…").get_by_text(
+                "Change Kernel…"
+            ).click()
+
+        self._set_environment_via_popup(kernel)
+
+    def clone_repo(self, git_url, branch=None, wait_for_completion=5):
+        """Clone a git repo into the jupyterlab file structure.
+
+        The terminal is a blackbox for the browser. We can't access any of the
+        displayed text, therefore we have no way of knowing if the commands
+        are done executing. For this reason, there is an unavoidable sleep
+        here that prevents playwright from moving on to ensure that the focus
+        remains on the Terminal until we are done issuing our commands.
+
+        Parameters
+        ----------
+        git_url: str
+            url of git repo (must be public, referenced as http rather than ssh
+            format, e.g. https://github.com/nebari-dev/nebari.git)
+        branch: bool or str
+            (Optional) branch to checkout from the repo. Defaults to None to use
+            the default branch.
+
+        Returns
+        -------
+        None
+        """
+        import datetime as dt
+
+        start = dt.datetime.now()
+
+        logger.debug(f">>> Clone git repo: {git_url}")
+
+        input_string = f"git clone {git_url}"
+        if branch:
+            input_string = f"{input_string} --branch {branch}"
+        self.page.get_by_text("File", exact=True).click()
+        self.page.get_by_text("New", exact=True).click()
+        self.page.get_by_role("menuitem", name="Terminal").get_by_text(
+            "Terminal"
+        ).click()
+        self.page.get_by_role("textbox", name="Terminal input").fill(input_string)
+        self.page.get_by_role("textbox", name="Terminal input").press("Enter")
+
+        # TODO temporary to check if environments are built
+        self.page.get_by_role("textbox", name="Terminal input").fill("conda env list")
+        self.page.get_by_role("textbox", name="Terminal input").press("Enter")
+
+        logger.debug(f"time to complete {dt.datetime.now() - start}")
+        # ensure that playwright doesn't move on/change context until all the
+        # above commands are complete.
+        time.sleep(wait_for_completion)
+        self.reset_workspace()
+
+
+if __name__ == "__main__":
+    import os
+
+    import dotenv
```

Comment:
Same as above. Why are those lazy imports?

Reply:
This one is intentional as its not needed for the remainder of the file, only when this is run as main.

Reply:
So is this runtime optimization? This was added as dev requirement and thus should always be available.

Reply:
yes, runtime optimization. Its added for ease-of-use when developing or running locally. On CI, the env vars are set in the GH Action yaml.

Reply:
Maybe just different opinions on the style here. We usually don't scope imports by usage inside a file, i.e. if an import is only used in one function, it still goes at the top of the file. The only reason I use lazy imports is if the package in question is an optional dependency that might not be available, but is not needed for most parts of the module. That is not the case here. `os` is from the standard library and `dotenv` is a dev requirement.

I would prefer to put them at the top, but I won't block over this. Feel free to resolve without fix if you disagree with me.